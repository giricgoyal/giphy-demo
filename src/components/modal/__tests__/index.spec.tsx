import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'

import Modal, { Props } from '../index'

describe('<Modal />', () => {
    let container
    const onCloseHandler = jest.fn()

    const defaultProps: Partial<Props> = {
        title: 'Some modal',
        onClose: onCloseHandler,
    }

    describe('When isVisible is false', () => {
        let classListRemoveSpy
        beforeEach(() => {
            classListRemoveSpy = jest.spyOn(document.body.classList, 'remove')
            ;({ container } = render(
                <Modal {...defaultProps}>
                    <div>I am Thor! God of thunder.</div>
                </Modal>,
            ))
        })

        test('should match snapshot', () => {
            expect(container).toMatchSnapshot()
        })

        test('should call the document.body.classList.add with overflow-hidden', () => {
            expect(classListRemoveSpy).toHaveBeenCalledWith('overflow-hidden')
        })
    })

    describe('When isVisible is true', () => {
        let classListAddSpy

        beforeEach(() => {
            classListAddSpy = jest.spyOn(document.body.classList, 'add')

            const props: Props = {
                ...defaultProps,
                isVisible: true,
            }

            ;({ container } = render(
                <Modal {...props}>
                    <div>I am Thor! God of thunder.</div>
                </Modal>,
            ))
        })

        test('should match snapshot', () => {
            expect(container).toMatchSnapshot()
        })

        test('should call the document.body.classList.remove with overflow-hidden', () => {
            expect(classListAddSpy).toHaveBeenCalledWith('overflow-hidden')
        })

        describe('onClose', () => {
            describe('When clicked', () => {
                beforeEach(() => {
                    fireEvent.click(screen.getByTestId('modal-close'))
                })

                test('Should call the onCloseHandler', () => {
                    expect(onCloseHandler).toHaveBeenCalledTimes(1)
                })
            })
        })
    })
})
