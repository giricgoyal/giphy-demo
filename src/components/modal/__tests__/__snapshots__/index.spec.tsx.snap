// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> When isVisible is false should match snapshot 1`] = `<div />`;

exports[`<Modal /> When isVisible is true should match snapshot 1`] = `
<div>
  <div
    class="modal-overlay"
    data-testid="modal-overlay"
  />
  <div
    class="modal"
    data-testid="modal"
  >
    <div
      class="modal__header"
    >
      <div
        class="modal__header__title"
      >
        Some modal
      </div>
      <svg
        class="modal__header__close icon"
        data-testid="modal-close"
      >
        <use
          xlink:href="#undefined"
        />
      </svg>
    </div>
    <div
      class="modal__body"
    >
      <div>
        I am Thor! God of thunder.
      </div>
    </div>
  </div>
</div>
`;
